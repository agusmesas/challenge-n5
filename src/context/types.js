export const productsTypes = {
  SET_INITIAL_CART: 'products/set-initial-cart',
  ADD_TO_CART: 'products/add-to-cart',
  BUY_CART: 'products/buy-cart',
  DELETE_CART_ITEM: 'products/delete-cart-item',
  SET_SELECTED_PRODUCT: 'products/set-selected-product',
  SET_PRODUCTS_LIST: 'products/set-products-list',
}

export const layoutTypes = {
  SET_LOADING: 'layout/set-loading'
}