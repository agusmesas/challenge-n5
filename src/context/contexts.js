'use client'
import { createContext } from 'react';

export const ProductsContext = createContext(null);
export const ProductsDispatchContext = createContext(null);

export const LayoutContext = createContext(null);
