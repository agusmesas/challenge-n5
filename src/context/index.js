import { useReducer } from 'react';
import { ProductsContext, ProductsDispatchContext, LayoutContext } from './contexts';

// Reducers
import productsReducer from './reducers/products';
import layoutReducer from './reducers/layout';

const initialState = {
  products: [],
  shoppingCart: []
}

const getInitialState = () => {
  const shoppingCart = JSON.parse(localStorage.getItem('cart')) || [];

  return {
    ...initialState,
    shoppingCart
  }
}

export default function Context({ children }) {
  const [productsState, productDispatch] = useReducer(
    productsReducer,
    initialState,
  );
  const [layoutState, layoutDispatch] = useReducer(
    layoutReducer,
    { loading: false },
  );

  function setInitialCart(cart) {
    productDispatch({
      type: 'set-initial-cart',
      cart,
    });
  }

  function addToCart(product) {
    productDispatch({
      type: 'add-to-cart',
      product,
    });
  }

  function buyCart(product) {
    productDispatch({
      type: 'buy-cart',
      product,
    });
  }

  function deleteCartItem(id) {
    productDispatch({
      type: 'delete-cart-item',
      id,
    });
  }

  function setSelectedProduct(product) {
    productDispatch({
      type: 'set-selected-product',
      product,
    });
  }

  function setProducts(products) {
    productDispatch({
      type: 'set-products-list',
      products
    });
  }

  function setLoading(loading) {
    layoutDispatch({
      type: 'set-loading',
      loading
    });
  }

  const productActions = {
    addToCart,
    buyCart,
    deleteCartItem,
    setSelectedProduct,
    setProducts,
    setInitialCart,
  };

  const layoutActions = {
    setLoading
  }

  return (
    <LayoutContext.Provider value={{ state: layoutState, actions: layoutActions }}>
      <ProductsContext.Provider value={productsState}>
        <ProductsDispatchContext.Provider value={productActions}>
          {children}
        </ProductsDispatchContext.Provider>
      </ProductsContext.Provider>
    </LayoutContext.Provider>
  );
}
