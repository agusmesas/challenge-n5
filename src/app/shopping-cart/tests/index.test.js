import '@testing-library/jest-dom';
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { ProductsContext, ProductsDispatchContext } from '../../../../src/context/contexts';
import Context from '../../../../src/context';
import Page from '../page';

jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      replace: jest.fn(),
    };
  }
}));


const renderComponent = () => {
  const actions = {
    addToCart: jest.fn(),
    buyCart: jest.fn(),
    deleteCartItem: jest.fn(), 
  }

  // const view = render(
  //   <ProductsContext.Provider value={{
  //     shoppingCart: [
  //       {
  //         "name": "Huevos",
  //         "price": 75000,
  //         "amount": 12,
  //         "id": 3,
  //         "quantity": 1,
  //       }]
  //     }}>
  //     <ProductsDispatchContext.Provider value={actions}>
  //       <Page />
  //     </ProductsDispatchContext.Provider>
  //   </ProductsContext.Provider>
  // );

  const view = render(
    <Context initialState={{
          products: [{
            "name": "Huevos",
            "price": 75000,
            "amount": 12,
            "id": 3,
            "quantity": 1,
          }],
          shoppingCart: [
            {
              "name": "Huevos",
              "price": 75000,
              "amount": 12,
              "id": 3,
              "quantity": 1,
            },
          ]
          }}>
      <Page />
    </Context>);

  return { view, actions };
}

describe('Shopping cart - Page' , () => {
  it('should render cart correctly', async () => {
    const { view } = renderComponent();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should buy products when click button', async () => {
    renderComponent();

    expect(screen.getByText('Total: $ 75000')).toBeInTheDocument();

    fireEvent.click(screen.getByRole('button', { name: 'Comprar' }));

    expect(screen.getByText('Total: $ 0')).toBeInTheDocument();
  });

  it('should delete product of cart', async () => {
    renderComponent();

    fireEvent.click(screen.getAllByRole('button')[2]);

    expect(screen.getByText('Total: $ 0')).toBeInTheDocument();
  });

  it('should add more quantity to cart', async () => {
    renderComponent();

    fireEvent.click(screen.getAllByRole('button')[1]);

    expect(screen.getByRole('textbox').value).toBe('2');
    expect(screen.getByText('Total: $ 150000')).toBeInTheDocument();
  });
})