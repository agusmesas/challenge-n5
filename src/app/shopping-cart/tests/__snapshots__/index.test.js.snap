// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping cart - Page should render cart correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="cart"
    >
      <h1>
        Carrito
      </h1>
      <div
        class="cart__products"
      >
        <div
          class="cart__products__item"
        >
          <p>
            Huevos - $ 75000 (12)
          </p>
          <div
            class="cart__products__item__actions"
          >
            <div>
              <div
                class="quantity"
              >
                <button
                  class="quantity__button quantity__button--decrement"
                >
                  â€”
                </button>
                <input
                  class="quantity__input"
                  type="text"
                  value="12"
                />
                <button
                  class="quantity__button quantity__button--increment"
                  disabled=""
                >
                  +
                </button>
              </div>
            </div>
            <button
              class="button button--secondary"
            >
              <svg
                class="icon undefined"
                fill="none"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 12V17"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  d="M14 12V17"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  d="M4 7H20"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  d="M6 10V18C6 19.6569 7.34315 21 9 21H15C16.6569 21 18 19.6569 18 18V10"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  d="M9 5C9 3.89543 9.89543 3 11 3H13C14.1046 3 15 3.89543 15 5V7H9V5Z"
                  stroke="#FFFFFF"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div>
        <p>
          Total: $ 900000
        </p>
        <button
          class="button button--primary"
        >
          Comprar
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
